Bugfix/54 - Implement 'hook end' event

DJ-Glock opened the pull request:
## Proposed changes
This pull request is created to address problem from the issue https://github.com/travisjeffery/mocha-teamcity-reporter/issues/54.
This change is a bit hacky for the case when hook failed because it looks like mocha does not fire `hook end` event if hook failed.

This change affects only a case when a report option `recordHookFailures` is set to true.

## Types of changes

What types of changes does your code introduce
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

DJ-Glock commented:
Hey @jamie-sherriff 
Can you please check this change before releasing new version?
All checks are green. TeamCity build is the same (11 failed tests as expected https://teamcity.jetbrains.com/viewLog.html?buildId=3437366&buildTypeId=TeamCityThirdPartyPlugins_MochaTeamcityReporter_Build).

The only thing I want to comment is this code:
```
		// Log testFinished for failed hook (hook end event is not fired for failed hook)
		if (recordHookFailures && !test.title.includes(ignoreHookWithName)) {
			if (test.title.includes(`"before all" hook`)
					|| test.title.includes(`"before each" hook`) 
					|| test.title.includes(`"after all" hook`)
					|| test.title.includes(`"after each" hook`)
				) {
				if(isNil(test.duration)){
					log(formatString(TEST_END_NO_DURATION, test.title, flowId));
				} else {
					log(formatString(TEST_END, test.title, test.duration.toString(), flowId));
				}
			}
		}
```
I put a comment to describe why I did it. Mocha does not fire `hook end` event if hook failed. So I had to add a condition to check if test name contains a hook name for case when hook failed. If it's not failed, it's logged as usual:
```
	runner.on('hook end', function (test) {
		if (recordHookFailures && !test.title.includes(ignoreHookWithName)) {
			if(isNil(test.duration)){
				log(formatString(TEST_END_NO_DURATION, test.title, flowId));
			} else {
				log(formatString(TEST_END, test.title, test.duration.toString(), flowId));
			}
		}
	});
```

As I checked output, it works fine. Functional tests were added (for `recordHookFailures`), regression was not changed.

jamie-sherriff commented on lib/teamcity.js:
```diff
@@ -116,6 +116,20 @@ function Teamcity(runner, options) {
 			}
 		}
+		// Log testFinished for failed hook (hook end event is not fired for failed hook)
+		if (recordHookFailures && !test.title.includes(ignoreHookWithName)) {
```
Hey @DJ-Glock can you double check this as I didn't pick this up in the previous PR but if you have a test title that includes the string undefined its going to evaluate to true because ignoreHookWithName is undefined when not set. 

Can you initialize this to false instead like the other reporter options?

e.g
`'testing for undefined'.includes(undefined)` //returns true

jamie-sherriff replied:
Might need a more comprehensive check than defaulting to false as includes also returns true on a false

`'testing for false'.includes(false)`

DJ-Glock replied:
Nice catch!

jamie-sherriff reviewed (changes requested):
Investigate test names that include undefined

DJ-Glock commented:
@jamie-sherriff 
I added conditions for undefined because boolean is not a good option here. I have also changed tests a little bit to cover this case:
```
	// Name of the function is an additional check for case when ignoreHookWithName is undefined
	beforeEach(function undefinedbeforeEachHookNoReporting() {
		throw new Error('Before each hook error fail');
	});
...
	// Name of the function is an additional check for case when ignoreHookWithName is undefined
	afterEach(function undefinedafterEachHook() {
		throw new Error('After each hook error fail');
	});
```

So finally it looks like this:

```
(ignoreHookWithName) ? ignoreHookWithName : null;
...
if (recordHookFailures && !ignoreHookWithName || recordHookFailures && ignoreHookWithName && !test.title.includes(ignoreHookWithName)) {
			log(formatString(TEST_START, test.title, flowId));
		}
```

Should be fine now. Are you okay with this solution?

jamie-sherriff reviewed (approved):
Looks great, thanks for fixing up and doing more testing.
